Add the @all alias

rgrinberg opened the pull request:
This alias includes the install, library artifacts, and executables in a
directory. As discussed #1220, this new alias will become the default alias.

rgrinberg committed:
Add the @all alias

This alias includes the install, library artifacts, and executables in a
directory

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

emillon reviewed (commented):
I think that corresponds to what we discussed in #1220. Can you add some docs? Also, do we need a new name for the current default?

rgrinberg commented:
@emillon Have another look please. I changed the default alias to `all` and things seem to be working. I did it in this PR as it was a rather simple change. The tests also seem to expose a bug in the variants implementations. That's a separate bug to be addressed in a subsequent PR.

Still need to update the change list and the manual.

rgrinberg commented:
One annoying thing about the current implementations is that user defined `@install` aliases will not be added to `@all`. It should be easy to implement, but I'm not sure which approach to take here. We can either:

* Just add some extra logic for compiling aliases to make sure `@all` stays up to date.
* Change `Build_system.Alias.add_alias_deps` to detect adding deps to `@install` and add them to `@all` as well.

rgrinberg committed:
Add tests for the all alias

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Make build foo/@all work correctly on @install stanzas

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Test to show that @all includes @install

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg commented:
A couple of other changes I've included:

* copy stanzas fall under `@all`
* install stanzas will now be included in `@subdir/all`.

rgrinberg commented:
I've also added .merlin files to the default alias.

Just as a note, I'm really liking this new behavior for the default alias. It has already helped me discover a couple of bugs with dune with regards to variants, and also detect some important (but not installed) stale artifacts in a couple of other projects. I'm pretty convinced that this is a much better default.

My only worry is that there is non trivial overhead to these extra rules. Hopefully it's not so big or we find a way to minimize it as I think this feature is worth it.

diml commented:
The new default seems better to me as well, however I still believe that such a change requires a major version bump. What about making this configurable via the `~/.config/dune/config` file? This way we can benefit from it immediately while still being conservative until dune 2.0.0.

rgrinberg commented:
> The new default seems better to me as well, however I still believe that such a change requires a major version bump. 

I see. Unfortunately, I think there are basically no benefits to this feature until we switch the default. The users who are confused by the current beahvior aren't users we can rely on tweaking defaults.

> What about making this configurable via the ~/.config/dune/config file? This way we can benefit from it immediately while still being conservative until dune 2.0.0.

We can just make the default for the `@default` alias configurable this way. I don't see a huge benefit in this option btw, we could always just leave `@all` around and try to encourage users to use it. At least until 2.0 comes around.

rgrinberg commented:
@diml @emillon I removed the commit which changed the default. We can revisit this when 2.0 comes around. The implementation is now ready for review.

diml commented:
It seems that we are calling `Lib_archives.make` twice and this function looks a bit expansive. What about setting up the alias in `install_rules.ml` directly?

rgrinberg commented:
Hmm, yeah that's a good point. Note that install_rules only calls this for public libraries however. So we'll still need the current code to do it for private libraries.

By the way, one nice thing about `@all` is that it's "better scoped" than `@install`. So for example, the following works as expected:

```
$ dune build @src/my-lib/all
```

I suppose it's not going to be an issue to setup `@all` in the correct directory for the public libraries. But if it will be necessary to add tricks like `on_load_dir` that would be really unfortunate.

Another point to consider: Would it make sense to define alias targets in sub directories? It would solve this particular problem and also with the more general problem of useless rules being eagerly created.

diml commented:
I suggest to bench a null build of `@install` on a big workspace such as the platform duniverse before changing the code.

rgrinberg commented:
I benchmarked this on the ocaml platform and there's no observable difference. It's around 1.6 seconds for a 0 build of `@install` with or without this PR.

diml commented:
Thanks for benching this. Let's not bother then

diml commented on src/lib_archives.mli:
```diff
@@ -0,0 +1,13 @@
+open Stdune
```
This is about a bit more than just library archives, what about `Lib_installed_files`?

rgrinberg replied:
Note that it doesn't in fact include all installed files. For example, it doesn't include the sources (anymore) or the ppx driver related stuff. We can also push the module specific stuff away from it and stick to archives only for `@all`. cmi's, cmt's, etc. will all be built transitively anyways.

diml replied:
that seems simpler indeed. And thiking about it more, if we really want to reference these files explicitely, we can also depend on all the files in the object directory

rgrinberg replied:
Okay, I've done that.

emillon commented on src/build_system.ml:
```diff
@@ -334,6 +335,7 @@ module Alias0 = struct
   let doc         = make "doc"
   let private_doc = make "doc-private"
   let lint        = make "lint"
+  let all         = make "all"
```
there's a bit of duplication between here and `is_standard`. What do you think about building a list ref in `make` (calling it `register` maybe?) and have `is_standard` implemented on top of it?

rgrinberg replied:
This bugs me as well. I'll do it in a separate PR if you don't mind as it has not relevance to this particular alias.

emillon replied:
absolutely.

emillon commented on src/lib_archives.mli:
```diff
@@ -0,0 +1,13 @@
+open Stdune
+
+type t =
```
I'm not against concrete types in general, but since this is a new one: would it make sense to start it abstract, based on how it is used?

rgrinberg replied:
I usually make the type abstract when the need arises, there's no harm in doing it here. I've done so.

emillon commented on test/blackbox-tests/test-cases/all-alias/user-defined/dune:
```diff
@@ -0,0 +1 @@
+(rule (with-stdout-to foo (run echo foobar)))
```
I think that you can use `write-file` here but that should get simplified to the same.

rgrinberg replied:
Actually, I'm trying to use an action that will show a trace when running `$ dune build @all --display short`. This is so that the test confirms the target was indeed built. This is a bit hacky of course, perhaps we should have a display mode that only includes the targets built.

emillon replied:
ah I didn't see it was `(run echo)`. that works!

rgrinberg committed:
Make Lib_archives abstract

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Move module specific artifacts back to Install_rules

Those will be build in the all alias b/c they're included in the archive anyway

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

diml commented:
Just to check, could you see what is the diff between the current dependencies of `@@dir/all` and `(glob_files dir/*)`? Just to make sure this PR is not a complicated implementation of the latter.

rgrinberg commented:
I'll double check, but it seems unlikely. Otherwise it would be a problem with any other alias as well.

diml commented:
What i meant is that if the two match, then we can use `(glob_files *)` as definition of the `all` alias instead of defining it manually.

rgrinberg commented:
It's quite similar. The only differences are:

* `(glob_files *)` includes all those js targets
* `(glob_files *)` includes the soon to be obsolete `.dune` target
* `@all` in the root dir of the context includes the all `@install` deps as well (META files)
* I haven't confirmed this, but I expect the `@@all` alias to include files requires for the the `install` stanza.

So indeed `@@all` is almost the same as `(glob_files *)`. The recursive behavior is quite important however. It could be analogous to `(glob_files **)` perhaps.

diml commented:
We get the recursive for free behavior by using `@all` though. Given your analysis, it seems to me that if we define the `all` alias in a directory as depending on all targets in this directory, then:
- we get the behavior we want
- this diff becomes much smaller
- it is simpler to explain what this alias does
- we do a bit less work inside dune

What do you think?

rgrinberg commented:
Yeah, I think this is a good idea. Will update the PR.

rgrinberg commented:
@diml one thing that this will require is to exclude the pesky `.js` targets. It's a shame that this isn't possible via re. I suppose this will PR will have to depend on the removal of those aliases first.

rgrinberg committed:
Add the * glob to the @all alias

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

diml commented:
> In retrospect, I wonder if it was simpler to handle js executables via modes instead of setting up the targets by default. It's quite unlikely that users ever want to create more than one of these modes: (best, js) for particular executables. And of course, it's something that we would still support by overriding the modes field.

yh, we should definitely have introduced the js mode right from the start. That's something to keep in mind for dune 2.0.0. In the meantime, we should just exclude these manually. We can't express it with re, but internally all we need is a `string -> bool` function, so we can just generalise the glob support.

rgrinberg committed:
Exclude .bc.js targets from all alias

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Remove manual adding of Lib_archives to alias

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg commented:
I've made the `@@all` alias depend on all the targets in the directory. I still kept the code that adds the the files in the install stanzas as dependencies of `@@all`. Should I remove that? Seems like it's a useful thing to verify - that all installable artifacts are built.

diml commented on src/gen_rules.ml:
```diff
@@ -37,23 +46,30 @@ module Gen(P : Install_rules.Params) = struct
-        (Some merlin, Some (lib.buildable.loc, cctx))
+        let cctx, merlin =
+          Lib_rules.rules lib ~dir ~scope ~dir_contents ~dir_kind in
+        (Some merlin, Some (lib.buildable.loc, cctx), None)
```
We are starting to return a lot of values here and it's becoming hard to remember what is what. What about introducing a record type?

rgrinberg replied:
Sure, I introduced a record type to clean this up. I think we do need a record with a field for every stanza of the same type eventually. We keep doing this binning everywhere.

diml commented:
> I've made the @@all alias depend on all the targets in the directory. I still kept the code that adds the the files in the install stanzas as dependencies of @@all. Should I remove that? Seems like it's a useful thing to verify - that all installable artifacts are built.

That seems good to me as well.

Please also add some doc in the manual.

rgrinberg committed:
Simplify per stanza accumulation

We need to extract merlins, compilation contexts, and js targets from all
stanzas in a dir. Use a record for this job.

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Add a section for built in aliases

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg commented:
I added a section for built in aliases to the manual as I don't think there was one before.

rgrinberg committed:
Update CHANGES for all alias

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

diml commented:
They are quickly describe in the "terminology" page, however the list is not exhaustive. We should replace this latter list by a pointer to the the new section.

diml commented on doc/usage.rst:
```diff
@@ -167,6 +167,26 @@ Is equivalent to:
+* ``lint`` - run linting tools.
+
+* ``all`` - build all available targets in a directory and installable artifacts
+  defined in that directory.
```
`default` is missing from this list

rgrinberg commented:
The failure is just the `output-obj` test.
